src/tracer.c

```diff
@@ -128,6 +128,39 @@ find(pid_t pid)
     return pinfo + i;
 }
 
+char *
+read_command_line(pid_t pid)
+{
+    char cmdline[64];
+    static char cmd[ARG_MAX];
+
+    sprintf(cmdline, "/proc/%d/cmdline", pid);
+
+    int fd = open(cmdline, O_RDONLY);
+
+    if (fd < 0) {
+	error(EXIT_FAILURE, errno, "failed to open file\n");
+    }
+
+    int bytes = read(fd, cmd, ARG_MAX);
+
+    if (bytes < 0) {
+	error(EXIT_FAILURE, errno, "failed to read file");
+    }
+
+    close(fd);
+
+    for (int i = 0; i < bytes; ++i) {
+	if (!cmd[i]) {
+	    cmd[i] = ' ';
+	}
```

Comment:
That's not enough.
Yes, `cat foo` is read like `cat\0foo\0` and you correctly transform it back.

But `cat "foo bar"` is read `cat\0foo bar\0` and you make it `cat foo bar`, which is not correct.

Reply:
@zvr Is this the reason why they are using nulls as separators? It kinda struck me as odd, yet another place where the mixed solution mentioned would really help us.

I guess i need some time to think about how to go about this.

Reply:
Not sure what you're trying to accomplish there -- especially with the handling of `"`...

Instead of doing two passes over the string, I'd suggest doing a single one to simply check whether there are spaces. If not, simply add the `arg`; if there are, add `"arg"`.

Yes, it won't work if the arg has both spaces and double quotes, but let's start with that. Otherwise, for a fool-proof solution, you have to look for double and single quotes, backslashes, etc.

Reply:
@zvr

Why don't we just... print it the way it is? With the null separators? It's honestly the simplest solution and all that's stopping us is the way record_triplet is implemented(that is, take a single null-terminated string as input).

After all we don't plan for the output to be human-readable. We could just add an extra `record` function that takes a string array(that is, to be more re-usable) or straight up go for a `record` function that takes a string of null-separated values.

This way we won't have to bother with all that stuff, accounting for special characters etc...

What's stopping us?

Reply:
If you try to `printf("%s, "cat\0foo bar\0")`, you will get `cat`(since the strings end at the first NULL character).

And we definitely want the RDF turtle file to be human-readable!

We _could_ use `"\0"` as separators, but in most cases spaces are so much more convenient.

Reply:
@zvr I am aware, that's why i mentioned that we need to alter the `record` function.

`cat` completely ignores \0 in my computer at least.

anyways so are you suggesting that we transform it from:

`gcc\0src.c\0-o\0test`

to:

`"gcc" "src.c" "-o" "test"` ?

Reply:
No, to
`"gcc src.c -o test"`

and if there is a space in an argument, use single quotes:
`"gcc 'My Programs/foo.c' -o foo"`

Reply:
@zvr Hmm so we either need to:

1. Pass the buffer once, so we count how many arguments `args_with_spaces` have a space in them and add `2 * args_with_spaces` to `bytes` for the extra quotes. Then allocate new `bytes + 2 * args_with_spaces` size of memory and start transforming the string.
2. Create a second buffer that has `bytes + 2 * argc`, that is assume that each and every `arg` has spaces in them. Kinda doubt this is the right solution because we still need a way to count `argc`, and i doubt we have one besides the obvious solution of passing the buffer and counting ourselves.
3. There is also the solution of creating a huge buffer. What's the worst case scenario here? `ARG_MAX` size of `' '` strings?

Reply:
@zvr I just noticed the way you implemented `record_envp`. Hadn't thought of this, why don't we implement `cmdline` this way too? Like:

`exec: <binary>`
`arg0: <args[0]>`
`arg1 <args[1]>`
.
.
.

Reply:
We can't, because the left-hand side on your example (which is the middle part of triples) have to be predefined.
In the environment case, they are all the same and order does not matter.

Anyway, I wrote #71 to record the command lines, so no need to add anything here.

Reply:
@zvr so do we close the pull request?

Reply:
Of course not -- this was about tracing exec's.
Rebase on the new main branch and add the corresponding code; simply do not add command-line arg processing.

Reply:
@zvr done